<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Construct a new gTrack — gTrack-class • gTrack</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Construct a new gTrack — gTrack-class"><meta property="og:description" content='S4 class for gTrack
Class gTrack defines a subsettable object that wraps formatting information around
genomic data (GRanges, GRangesList, RleLists, UCSC formats, ffTrack) and can be displayed
in a "genome browser" style plot.
Arguments described as "formatting" are vectors.  They are replicated (if necessary) to match the length of the object.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gTrack</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="http://github.com/mskilab/gTrack/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct a new <code>gTrack</code></h1>
    <small class="dont-index">Source: <a href="http://github.com/mskilab/gTrack/blob/HEAD/R/gTrack.R" class="external-link"><code>R/gTrack.R</code></a></small>
    <div class="hidden name"><code>gTrack-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>S4 class for <code>gTrack</code></p>
<p>Class <code>gTrack</code> defines a subsettable object that wraps formatting information around
genomic data (GRanges, GRangesList, RleLists, UCSC formats, ffTrack) and can be displayed
in a "genome browser" style plot.</p>
<p>Arguments described as "formatting" are vectors.  They are replicated (if necessary) to match the length of the object.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gTrack</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  y.field <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  mdata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  edges <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vars <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colormaps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  ygap <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">""</span>,</span>
<span>  stack.gap <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  cex.label <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  gr.cex.label <span class="op">=</span> <span class="va">cex.label</span> <span class="op">*</span> <span class="fl">0.8</span>,</span>
<span>  gr.srt.label <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  col <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  border <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  angle <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  name <span class="op">=</span> <span class="st">""</span>,</span>
<span>  gr.colorfield <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  y.quantile <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  y.cap <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  lwd.border <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  hadj.label <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  vadj.label <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  round <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  ywid <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  ypad <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  seqinfo <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  circles <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  lines <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bars <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  draw.paths <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  path.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  path.lwd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  draw.var <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  triangle <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">mdata</span><span class="op">)</span>,</span>
<span>  max.ranges <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>  source.file.chrsub <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  y0.bar <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  yaxis <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">y.field</span><span class="op">)</span>,</span>
<span>  yaxis.pretty <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  chr.sub <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  edgevars <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  gr.labelfield <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  grl.labelfield <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  grl.colorfield <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  grl.cexfield <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  xaxis.prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  xaxis.unit <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xaxis.suffix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  xaxis.round <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  xaxis.cex.label <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xaxis.newline <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  xaxis.chronly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  xaxis.width <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  xaxis.interval <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  xaxis.label.angle <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  xaxis.ticklen <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xaxis.cex.tick <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  sep.lty <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  sep.lwd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  sep.bg.col <span class="op">=</span> <span class="st">"gray95"</span>,</span>
<span>  sep.draw <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  y0 <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  y1 <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  m.sep.lwd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  cmap.min <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  cmap.max <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  labels.suppress <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  labels.suppress.grl <span class="op">=</span> <span class="va">labels.suppress</span>,</span>
<span>  labels.suppress.gr <span class="op">=</span> <span class="va">labels.suppress</span>,</span>
<span>  bg.col <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  formatting <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>Instance of one of the following objects: (1) GRanges (2) GRangesList (3) ffTrack or
(4) character representing Bed, Wig, BigWig, or .rds GRanges file.  It can also be a list of the above (if specifying multiplying
 tracks.</p></dd>


<dt>y.field</dt>
<dd><p>vector or scalar character referencing meta data field of GRanges or GRangesList to use for "y axis" coordinate when plotting
numeric tracks, (note: for a RleList or ffTrack this field is automatically set to "score"), default is NA for non-numeric tracks</p></dd>


<dt>mdata</dt>
<dd><p>Square matrix representing values at connections between genomic elements (e.g. chr8:1,000-2,000 to chr3:5,000-6,000). The number of
rows (and thus columns) must be equal to the length of the <code>data</code> element. The value of the (i,j) element of this matrix represents a value
for the intersection of <code>data[i]</code> and <code>data[j]</code>. Note that all connections are assumed to be symmetric, and only the upper triangle
will be read. Displays a <code>warning</code> if the matrix is non-symmetric</p></dd>


<dt>edges</dt>
<dd><p>Data frame of columns $from, $to, and optional fields $col, $lwd, and $lty, specifying edges linking data items.
Also can be a list of the above if specifying multiple tracks (and must be compatible in length with data arg)</p></dd>


<dt>colormaps</dt>
<dd><p>length(.Object) length named list of named vectors whose entry i maps uniques value of a data field to colors.  The data.field is specified by the name of the list entry, and the unique values / colors are specified by the named vector.</p></dd>


<dt>height</dt>
<dd><p>vector or scalar numeric specifying height of track(s) (in relative units)</p></dd>


<dt>ygap</dt>
<dd><p>vector or scalar numeric specifying gap between tracks</p></dd>


<dt>stack.gap</dt>
<dd><p>vector or scalar.numeric specifying x gap between stacking non-numeric GRanges or GrangesLists items in track(s)</p></dd>


<dt>cex.label</dt>
<dd><p>vector or scalar numeric specifying the expansion factor of the range labels (formatting)</p></dd>


<dt>gr.cex.label</dt>
<dd><p>vector or scalar numeric specifying GRanges label character expansion (default is cex.label)  (formatting)</p></dd>


<dt>gr.srt.label</dt>
<dd><p>vector or scalar numeric between 0 and 180 specifying rotation of GRanges labels  (formatting)</p></dd>


<dt>col</dt>
<dd><p>vector or scalar character specifying static color for track(s), if NA then color is specified by colormaps() property or gr.colorfield or col meta data field of GRanges / GRangesList data object</p></dd>


<dt>border</dt>
<dd><p>vector or scalar character specifying static border for polygons in track(s), if NA then $border is determine duing gr.colorfield / colormap or meta field $border of GRanges / GRangesList</p></dd>


<dt>angle</dt>
<dd><p>vector of scalar numeric specifying angle of polygons that represent signed range'
items (only relevant if used within chainedTracks object)</p></dd>


<dt>name</dt>
<dd><p>vector or scalar character specifying name of this track, which will be displayed on label to the left of the track</p></dd>


<dt>lwd.border</dt>
<dd><p>vector or scalar integer specifying the thickness of the polygon borders (formatting)</p></dd>


<dt>hadj.label</dt>
<dd><p>vector or scalar numeric specifying the horizontal adjustment of the range labels (formatting)</p></dd>


<dt>vadj.label</dt>
<dd><p>vector or scalar numeric specifying the vertical adjustment of the range labels (formatting)</p></dd>


<dt>ywid</dt>
<dd><p>vector or scalar numeric specifying the y-extent of individual ranges (in local plot coordinates)</p></dd>


<dt>ypad</dt>
<dd><p>vector or scalar numeric between 0 and 1 specifying how much whitespace padding to add within panel (formatting)</p></dd>


<dt>circles</dt>
<dd><p>vector or scalar logical specifying whether to scatter plot range data (formatting)</p></dd>


<dt>lines</dt>
<dd><p>vector or scalar logical specifying whether to line plot range data (formatting)</p></dd>


<dt>bars</dt>
<dd><p>vector or scalar logical specifying whether to bar plot range data (formatting)</p></dd>


<dt>draw.paths</dt>
<dd><p>vector or scalar logical specifying whether to interpret GRangesLists as "paths" and connect them with a
a set of spline curves.  (formatting)</p></dd>


<dt>path.col</dt>
<dd><p>vector or scalar character specifying color of path (only applicable for tracks in which draw.paths = T) (formatting)</p></dd>


<dt>draw.var</dt>
<dd><p>vector or scalar logical specifying whether to draw.var for GRanges / GRangesList  specifying reads
(GRanges must contain $cigar +/- $MD field) (formatting)</p></dd>


<dt>max.ranges</dt>
<dd><p>vector or scalar numeric specifying what is the max number of ranges to plot in a window (formatting)</p></dd>


<dt>source.file.chrsub</dt>
<dd><p>vector or scalar logical specifying whether or not sub "chr" out of any external files (e.g. UCSC style files) (formatting)</p></dd>


<dt>y0.bar</dt>
<dd><p>vector or scalar numeric specifying where to draw the lower boundary of a bar in a bar plot (only applicable if bars == T) (formatting)</p></dd>


<dt>yaxis</dt>
<dd><p>vector or scalar logical specifying whether to print yaxis (formatting)</p></dd>


<dt>yaxis.pretty</dt>
<dd><p>vector or scalar positive integer specifying how many ticks to optimally draw on yaxis (formatting)</p></dd>


<dt>gr.labelfield</dt>
<dd><p>vector or scalar character specifying which GRanges meta data field to use for GRanges label (default "label") (formatting)</p></dd>


<dt>grl.labelfield</dt>
<dd><p>vector or scalar character specifying which GRanges meta data field to use for GRangesList label (default "label") (formatting)</p></dd>


<dt>grl.colorfield</dt>
<dd><p>vector or scalar character specifying which GRanges meta data field to use for GRangesList labels color (default "label.color") (formatting)</p></dd>


<dt>grl.cexfield</dt>
<dd><p>vector or scalar character specifying which GRanges meta data field to use for GRangesList labels cex (default "cex.label") (formatting)</p></dd>


<dt>xaxis.unit</dt>
<dd><p>vector or scalar numeric specifying the unit that will be used in describing x axis coordinates (TODO: move to display)  (formatting)</p></dd>


<dt>xaxis.suffix</dt>
<dd><p>vector or scalar numeric specifying the suffix that will be used in describin x axis coordinates (TODO: move to display) (formatting)</p></dd>


<dt>xaxis.round</dt>
<dd><p>vector or scalar non-neg integer specifying number of decimals to round xaxis coordinate labels (formatting)</p></dd>


<dt>xaxis.newline</dt>
<dd><p>vector or scalar logical specifying whether to draw a newline in the xaxis coordinate labels (formatting)</p></dd>


<dt>xaxis.width</dt>
<dd><p>Logical scalar specifying whether to add window width to xaxis window labels [TRUE]</p></dd>


<dt>xaxis.label.angle</dt>
<dd><p>vector or scalar numeric between 0 and 360 specifying angle with which to draw xaxis coordinate labels (formatting)</p></dd>


<dt>xaxis.ticklen</dt>
<dd><p>Scalar numeric specifying lengths for axis ticks  (formatting)</p></dd>


<dt>xaxis.cex.tick</dt>
<dd><p>Scalar numeric specifying expansion factor for axis tick labels  (formatting)</p></dd>


<dt>sep.lty</dt>
<dd><p>Scalar integer specifying line style for window separators [2] (dashed line)</p></dd>


<dt>sep.lwd</dt>
<dd><p>Scalar numeric specifying line thickness for window separators [1]</p></dd>


<dt>sep.bg.col</dt>
<dd><p>Color (supplied by name or hex code) for the background of the windows [gray95]</p></dd>


<dt>sep.draw</dt>
<dd><p>Logical allowing separators to be turned off [FALSE]</p></dd>


<dt>cmap.min</dt>
<dd><p>minimum saturating data value of color map for triangle plot</p></dd>


<dt>cmap.max</dt>
<dd><p>maximum saturating data value of color map for triangle plot</p></dd>


<dt>labels.suppress</dt>
<dd><p>whether to suppress labels (both GRangesList and GRanges)</p></dd>


<dt>labels.suppress.gr</dt>
<dd><p>whether to suppress GRangesList labels</p></dd>


<dt>legend.maxitems</dt>
<dd><p>Scalar positive integer specifying what is the maximum number of items to include in legend [Inf]</p></dd>


<dt>label.suppress</dt>
<dd><p>vector or scalar logical flag specifying whether to suppress all GRanges / GRangesList label drawing  (formatting)</p></dd>


<dt>label.suppress.gr</dt>
<dd><p>vector or scalar logical flag specifying whether to suppress GRanges label drawing  (formatting)</p></dd>


<dt>label.suppress.grl</dt>
<dd><p>vector or scalar logical flag specifying whether to suppress GRangesList label drawing  (formatting)</p></dd>


<dt>cex.label.gr</dt>
<dd><p>size of bottom level label (i.e. GRanges in a GrangesListitem, e.g. exons in a gene)</p></dd>


<dt>split</dt>
<dd><p>vector or scalar logical flag specifying whether to split when lifting (only relevant if used wihtin chainedTracks object)</p></dd>


<dt>xaxis.nticks</dt>
<dd><p>vector or scalar positive integer specifying how many xaxis ticks to optimally draw (formatting)</p></dd>


<dt>y.grid.col</dt>
<dd><p>vector or scalar character specifying color of "gridlines" used to specify numeric track data (formatting)</p></dd>


<dt>y.grid.cex</dt>
<dd><p>vector or scalar non-neg numeric specifying character expansion for y tick / y grid labels (formatting)</p></dd>


<dt>y.grid.lty</dt>
<dd><p>vector or scalar positive integer specifying line style of y grid lines for numeric tracks (formatting)</p></dd>


<dt>y.grid.lwd</dt>
<dd><p>vector or scalar positive integer specifying thickness of y grid lines for numeric tracks (formatting)</p></dd>


<dt>y.grid.labx</dt>
<dd><p>vector or scalar positive integer specifying fraction of xlim left of plot to place y axis labels (formatting)</p></dd>


<dt>path.col.arrow</dt>
<dd><p>vector or scalar character specifying color of arrow of path (only applicable for tracks in which draw.paths = T) (formatting)</p></dd>


<dt>path.cex.arrow</dt>
<dd><p>vector or scalar numeric &gt; 0 specifying expansion factor of arrow of path (only applicable for tracks
in which draw.paths = T)  (formatting)</p></dd>


<dt>path.stack.y.gap</dt>
<dd><p>vector or scalar numeric &gt; 0 specifying y stack gap of paths (only applicable for tracks in which draw.paths = T)  (formatting)</p></dd>


<dt>path.stack.x.gap</dt>
<dd><p>vector or scalar numeric &gt; 0 specifying x stack gap for paths  (only applicable for tracks in which draw.paths = T)  (formatting)</p></dd>


<dt>path.cex.v</dt>
<dd><p>vector or scalar numeric &gt; 0 specifying vertical bulge of sline in paths (only applicable for tracks in which draw.paths = T)  (formatting)</p></dd>


<dt>path.cex.h</dt>
<dd><p>vector or scalar numeric &gt; 0 specifying horizontal bulge of spline in paths (formatting)
(only applicable for tracks in which draw.paths = T)  (formatting)</p></dd>


<dt>draw.backbone</dt>
<dd><p>vector or scalar logical specifying whether to draw "backbone" connecting different items in a GRangesList item (formatting)</p></dd>


<dt>...</dt>
<dd><p>additional formatting arguments to gTrack</p></dd>

</dl></div>
    <div id="slots">
    <h2>Slots</h2>
    

<dl><dt>data</dt>
<dd><p>length(.Object) length list containing genomic data (e.g. GRanges, GrangesLists, RleLists, path to ucsc file or ffTrack file on disk)</p></dd>

  <dt>seqinfo</dt>
<dd><p>Seqinfo object</p></dd>

  <dt>colormap</dt>
<dd><p>length(.Object) length named list of named vectors whose entry i specifies the colormap for the meta data field of object entry i (specified by the name) and maps unique value of that data field to colors (specified by the named vector)</p></dd>

  <dt>edges</dt>
<dd><p>list of data.frames of length length(.Object) which has columns $from, $to, and optional fields $col, $lwd, and $lty to specify splined edges joining data items in the corresponding track</p></dd>

  <dt>formatting</dt>
<dd><p><code>data.frame</code> holding all of the formatting options</p></dd>

  <dt>mdata</dt>
<dd><p><code>matrix</code> holding interaction data between genomic loci (e.g. hiC)</p></dd>

  <dt>vars</dt>
<dd><p>List of variants associated with track</p></dd>


</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcin Imielinski</p>
<p>Marcin Imielinski</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jeremiah Wala, Marcin Imielinski.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

